temporalcli/commands.workflow.go

```diff
@@ -187,33 +186,61 @@ func (c *TemporalWorkflowTerminateCommand) run(cctx *CommandContext, _ []string)
 	return nil
 }
 
-func (c *TemporalWorkflowUpdateCommand) run(cctx *CommandContext, args []string) error {
-	cl, err := c.Parent.ClientOptions.dialClient(cctx)
+func (c *TemporalWorkflowUpdateStartCommand) run(cctx *CommandContext, args []string) error {
+	waitForStage := client.WorkflowUpdateStageUnspecified
+	switch c.WaitForStage.Value {
+	case "accepted":
+		waitForStage = client.WorkflowUpdateStageAccepted
+	}
+	if waitForStage != client.WorkflowUpdateStageAccepted {
+		return fmt.Errorf("invalid wait for stage: %v, valid values are: 'accepted'", c.WaitForStage)
+	}
+	return workflowUpdateHelper(cctx, c.Parent.Parent.ClientOptions, c.PayloadInputOptions, c.UpdateOptions, waitForStage)
+}
+
+func (c *TemporalWorkflowUpdateExecuteCommand) run(cctx *CommandContext, args []string) error {
+	return workflowUpdateHelper(cctx, c.Parent.Parent.ClientOptions, c.PayloadInputOptions, c.UpdateOptions, client.WorkflowUpdateStageCompleted)
+}
+
+func workflowUpdateHelper(cctx *CommandContext,
+	clientOpts ClientOptions,
+	inputOpts PayloadInputOptions,
+	updateOpts UpdateOptions,
+	waitForStage client.WorkflowUpdateStage,
+) error {
+	cl, err := clientOpts.dialClient(cctx)
 	if err != nil {
 		return err
 	}
 	defer cl.Close()
 
-	// Get raw input
-	input, err := c.buildRawInput()
+	input, err := inputOpts.buildRawInput()
 	if err != nil {
 		return err
 	}
 
 	request := client.UpdateWorkflowOptions{
-		WorkflowID:          c.WorkflowId,
-		RunID:               c.RunId,
-		UpdateName:          c.Name,
-		UpdateID:            c.UpdateId,
-		FirstExecutionRunID: c.FirstExecutionRunId,
+		WorkflowID:          updateOpts.WorkflowId,
+		RunID:               updateOpts.RunId,
+		UpdateName:          updateOpts.Name,
+		UpdateID:            updateOpts.UpdateId,
+		FirstExecutionRunID: updateOpts.FirstExecutionRunId,
 		Args:                input,
-		WaitForStage:        client.WorkflowUpdateStageCompleted,
+		WaitForStage:        waitForStage,
 	}
 
 	updateHandle, err := cl.UpdateWorkflow(cctx, request)
 	if err != nil {
 		return fmt.Errorf("unable to update workflow: %w", err)
 	}
+	if waitForStage == client.WorkflowUpdateStageAccepted {
+		return cctx.Printer.PrintStructured(
+			struct {
+				Name     string `json:"name"`
+				UpdateID string `json:"updateId"`
+			}{Name: updateOpts.Name, UpdateID: updateHandle.UpdateID()},
+			printer.StructuredOptions{})
+	}
 
 	var valuePtr interface{}
 	err = updateHandle.Get(cctx, &valuePtr)
```

Comment:
Note, this update error may contain lots of valuable information beyond just the message appended to "unable to update workflow: ". It may have details, causes, etc. Don't need to solve now (not solved for query either), but we should probably open an issue to provide a better dump of failure upon failure. I think for workflow execute we dump the entire close event.

Reply:
https://github.com/temporalio/cli/issues/651